test/extension/handlers/topologymutation/handler.go

```diff
@@ -19,63 +19,225 @@ package topologymutation
+	}
+
+	// 2) If the Kubernetes version from builtin.controlPlane.version is below 1.24.0 set "cgroup-driver": "cgroupfs" to
+	//    - kubeadmConfigSpec.InitConfiguration.KubeletExtraArgs
+	//    - kubeadmConfigSpec.JoinConfiguration.KubeletExtraArgs
+	maximumVersionForPatch, err := version.ParseMajorMinorPatchTolerant("v1.24.0")
+	if err != nil {
+		return errors.Wrap(err, "could not set parse kubernetes version v1.24.0")
+	}
+	_, cpVersion, err := getStringValueForVariable(templateVariables, "builtin.controlPlane.version")
+	if err != nil {
+		return errors.Wrap(err, "could not set cgroup-driver to control plane template kubeletExtraArgs")
+	}
+	controlPlaneVersion, err := version.ParseMajorMinorPatchTolerant(cpVersion)
 	if err != nil {
 		return err
 	}
+	if version.Compare(controlPlaneVersion, maximumVersionForPatch) == -1 {
+		log.Info("Setting kubeadmControlPlaneTemplate cgroup-driver", "cgroup-driver", cgroupfs)
+		kubeadmControlPlaneTemplate.Spec.Template.Spec.KubeadmConfigSpec.InitConfiguration.NodeRegistration.KubeletExtraArgs["cgroup-driver"] = cgroupfs
+		kubeadmControlPlaneTemplate.Spec.Template.Spec.KubeadmConfigSpec.JoinConfiguration.NodeRegistration.KubeletExtraArgs["cgroup-driver"] = cgroupfs
+	}
 
+	// 3) Patch RolloutStrategy RollingUpdate MaxSurge with the value from the Cluster Topology variables.
+	//    If this is unset continue as this variable is not required.
+	if found, kcpControlPlaneMaxSurge, err := getStringValueForVariable(templateVariables, "kubeadmControlPlaneMaxSurge"); found {
+		if err != nil {
+			return errors.Wrap(err, "could not set spec.template.spec.rolloutStrategy.rollingUpdate.maxSurge")
+		}
+		// This has to be converted to IntOrString type.
+		kubeadmControlPlaneMaxSurgeIntOrString := intstrutil.Parse(kcpControlPlaneMaxSurge)
+		log.Info("Setting kubeadmControlPlaneMaxSurge", "kubeadmControlPlaneMaxSurge", kubeadmControlPlaneMaxSurgeIntOrString.Type)
+		if kubeadmControlPlaneTemplate.Spec.Template.Spec.RolloutStrategy == nil {
+			kubeadmControlPlaneTemplate.Spec.Template.Spec.RolloutStrategy = &controlplanev1.RolloutStrategy{}
+		}
+		if kubeadmControlPlaneTemplate.Spec.Template.Spec.RolloutStrategy.RollingUpdate == nil {
+			kubeadmControlPlaneTemplate.Spec.Template.Spec.RolloutStrategy.RollingUpdate = &controlplanev1.RollingUpdate{}
+		}
+		kubeadmControlPlaneTemplate.Spec.Template.Spec.RolloutStrategy.RollingUpdate.MaxSurge = &kubeadmControlPlaneMaxSurgeIntOrString
+	}
+	return nil
+}
+
+// patchDockerClusterTemplate patches the DockerClusterTemplate.
+func patchDockerClusterTemplate(_ context.Context, dockerClusterTemplate *infrav1.DockerClusterTemplate, templateVariables map[string]apiextensionsv1.JSON) error {
 	// Unquote the JSON string.
-	stringValue, err := strconv.Unquote(string(value.Raw))
+	// patch lb
+	found, lbImageRepo, err := getStringValueForVariable(templateVariables, "lbImageRepository")
```

Comment:
probably because of WIP. skip on !found?

Reply:
The implementation here is for a "required" variable, so the !found case isn't important IMO.

Reply:
On one side yes on the other side I would see it as best practice that the patch implementation handles cases like this well.

(and setting lbImageRepository should be optional)

Reply:
Setting lbImageRepository is a required variable in the current version - should I change it in this implementation?
https://github.com/kubernetes-sigs/cluster-api/blob/2c2a1f1920dfb4f56b2d582aaad2a21c22140382/test/e2e/data/infrastructure-docker/v1beta1/clusterclass-quick-start-runtimesdk.yaml#L36

Reply:
Yes

Reply:
I think it's good to match the pattern we want to propagate of a extension which can handle a bunch of patches at the same time, while users / ClusterClass authors can decide to use them or not (but are not forced to use them, except that's absolutely necessary)